tests/func/test_add.py

```diff
@@ -190,6 +190,38 @@ def test_add_file_in_dir(tmp_dir, dvc):
     assert stage.outs[0].def_path == "subdata"
 
 
+def test_add_filtered_files_in_dir(tmp_dir, dvc):
+    tmp_dir.gen(
+        {
+            "dir": {
+                "subdir": {
+                    "subdata": "subdata content",
+                    "esubdata": "extra subdata content",
+                    "subdata123": "subdata content 123",
+                }
+            }
+        }
+    )
+    subdir_path = os.path.join("dir", "subdir", "subdata*")
```

Comment:
Should we test more features or just single wildcards?

Reply:
I'll add some more use cases for shell-style wildcards, and possible some negative ones for recursive matching which will not work as it is now, unless we enable the recursive flag. do we want that ?

Reply:
@ju0gri Note that the more the tests - the more expensive it get's, as we run them in real filesystem. I suppose os.glob has some spec that it sticks to, so it probably doesn't make sense to test it all.

Recursive flag is for double wildcards only, right? Definitely seems useful. Any reason why it is disabled by default? Maybe some perf issues (except for the obvious misuse).

Reply:
Yes, the recursive flag enables `**`. According to their docs: 
> Using the “**” pattern in large directory trees may consume an inordinate amount of time.

Reply:
Would adding one tests for each use case defined in `fnmatch` manual be an acceptable compromise ?
```
* | matches everything
? | matches any single character
[seq] | matches any character in seq
[!seq] | matches any character not in seq
** | searches recursively
```

Reply:
> Yes, the recursive flag enables **. According to their docs:

Sure, I am wondering whether it has some hidden cons. E.g. maybe it is considerably slower by itself when that pattern is supported. I guess need to give it a shot/look into internals for that.

Reply:
> Would adding one test for each use case defined in fnmatch manual be an acceptable compromise ?

Sounds good.

Btw, pathspec might not be needed in some cases. E.g. `dvc add '?*[seq]**'`, where all the special stuff is already shieled by the single quotes. Current patch enabled globbing by default, but it looks like we should probably make it optional.

Reply:
Yes, makes sense. I am bad with naming, but added an argument called `pathpattern` for the `add` command to enable globing of the passed in target. Also, as I am not very familiar with the structure of the repo, are there other places where I need to add coverage for the new argument ?